contracts/IPAccountImpl.sol

```diff
@@ -59,15 +59,26 @@ contract IPAccountImpl is ERC6551, IPAccountStorage, IIPAccount {
         return super.token();
     }
 
-    /// @notice Checks if the signer is valid for the given data
+    /// @notice Checks if the signer is valid for executing specific actions on behalf of the IP Account.
     /// @param signer The signer to check
-    /// @param data The data to check against
-    /// @return The function selector if the signer is valid, 0 otherwise
+    /// @param data The data to be checked. The data should be encoded as `abi.encode(address to, bytes calldata)`,
+    /// where `address to` is the recipient and `bytes calldata` is the calldata passed to the recipient.
+    /// If `data.length == 0`, it is also considered valid, implying that the signer is valid for all actions.
+    /// @return result The function selector if the signer is valid, 0 otherwise
     function isValidSigner(
         address signer,
         bytes calldata data
-    ) public view override(ERC6551, IIPAccount) returns (bytes4) {
-        return super.isValidSigner(signer, data);
+    ) public view override(ERC6551, IIPAccount) returns (bytes4 result) {
+        result = bytes4(0);
+        address to = address(0);
+        bytes memory callData = "";
+        if (data.length > 0) {
+            if (data.length < 32) revert Errors.IPAccount__InvalidCalldata();
+            (to, callData) = abi.decode(data, (address, bytes));
+        }
+        if (this.isValidSigner(signer, to, callData)) {
```

Comment:
FYI, this "if" wrap is redundant as it will produce the same result as just calling `this.isValidSigner(signer, to, callData)`. Since `checkPermission` is a reverting function, `isValidSigner` simply reverts in the case of invalid permission.